recipes/octave/meta.yaml

```diff
@@ -0,0 +1,104 @@
+{% set version = "4.2.1" %}
+{% set variant = "openblas" %}
+
+package:
+  name: octave
+  version: {{ version }}
+
+source:
+  url: https://ftp.gnu.org/gnu/octave/octave-{{ version }}.tar.gz
+  sha256: 80c28f6398576b50faca0e602defb9598d6f7308b0903724442c2a35a605333b
+  patches:
+    # patches taken from macports
+    # https://github.com/macports/macports-ports/tree/master/math/octave/files
+    # Link fails with undefined symbol ~file_stat
+    - patch-file-stat.cc.diff
+    # qscintilla2 library is not found without this
+    - patch-qscintilla2.diff
+
+build:
+  number: 0
+  skip: True  # [win]
+  features:
+    - blas_{{ variant }}
+
+requirements:
+  build:
+    - toolchain
+    - gcc
+    - sed
+    - perl 5.20.3.1
+    - openblas 0.2.19|0.2.19.*
+    - blas 1.1 {{ variant }}
+    - pcre
```

Comment:
A good reference for backward compatibility is https://abi-laboratory.pro/tracker/
Basically you will want to ```build``` with one version and then allow any backward compatible version as a ```run``` requirement.

For example, for [libsndfile](https://abi-laboratory.pro/tracker/timeline/libsndfile/) we have this:

```
1.0.28 	2017-04-02 	1 	changelog 	100% 	0 	0
1.0.27 	2016-06-19 	1 	changelog 	100% 	0 	0
1.0.26 	2015-11-22 	1 	changelog 	100% 	6 new 	0
1.0.25 	2011-07-13 	1 	changelog 	100% 	0 	0
1.0.24 	2011-03-23 	1 	changelog 	100% 	0 	0
1.0.23 	2010-10-10 	1 	changelog 	100% 	0 	0
```

All 1.0.x versions are backward compatible with 1.0.(x-1). But note that some versions add symbols. This is important. That is why you sometimes need to be more specific about build requirements. For example, this would be ok:

```
  build:
    - libsndfile 1.0.25
  run
    - libsndfile >=1.0.25,<1.1
```

In the case above any libsndfile >=1.0.25 will contain the symbols present on 1.0.25. But this could cause problems:

```
  build:
    - libsndfile 1.0.*
  run
    - libsndfile 1.0.*
```

This would allow the package to be built against 1.0.26 (which has 6 new symbols) and run with 1.0.25 (which does not have the 6 new symbols).
Of course it is possible that a future version 1.0.29 breaks backward compatibility, but given the libsndfile history it seems unlikely.

So in summary, on build requirements you should try to use a pin that only allows version that are a replacement for each other (they cannot add/remove symbols or change interfaces). For example, ```xz 5.2.*``` as a build requirement is fine because 5.2.1, 5.2.2, 5.2.3 does not add/remove anything when compared to 5.2.0: https://abi-laboratory.pro/tracker/timeline/xz/

And for the run requirements you can be more loose and allow versions that add things when compared with the version used during build.

When there is no info about some library or when the library tends to break compatibility very often I would recommend being more strict and pin the whole version. Better having conda complain about unmet dependencies than getting undefined symbols or crashes.

I hope this was not too confusing ;)